components/GraphPane/colorizers/ModuleTypeColorizer.tsx

```diff
@@ -12,14 +14,94 @@ export default {
   legend() {
     return (
       <>
-        <LegendColor color={COLORIZE_MODULE_CJS}>CommonJS (CJS)</LegendColor>
-        <LegendColor color={COLORIZE_MODULE_ESM}>EcmaScript (ESM)</LegendColor>
+        <LegendColor color={COLORIZE_MODULE_CJS}>CJS Only</LegendColor>
+        <LegendColor color={COLORIZE_MODULE_DUAL}>CJS and ESM</LegendColor>
+        <LegendColor color={COLORIZE_MODULE_ESM}>ESM Only</LegendColor>
+        <LegendColor color={COLORIZE_MODULE_TYPES}>TS Types</LegendColor>
       </>
     );
   },
 
   async colorForModule(module: Module) {
-    const pkg = module.package as PackageJSON;
-    return pkg.type === 'module' ? COLORIZE_MODULE_ESM : COLORIZE_MODULE_CJS;
+    const pkgType = detectPackageType(module.package as PackageJSON);
+    if (pkgType.esm && pkgType.cjs) {
+      return COLORIZE_MODULE_DUAL;
+    }
+    if (pkgType.esm) {
+      return COLORIZE_MODULE_ESM;
+    }
+
+    if (pkgType.cjs) {
+      return COLORIZE_MODULE_CJS;
+    }
+
+    if (pkgType.types) {
+      return COLORIZE_MODULE_TYPES;
+    }
   },
 };
+
+type PackageModuleType = {
+  esm: boolean;
+  cjs: boolean;
+  types: boolean;
+};
+
+// Package-type detection logic
+//
+// Logic to look at a PackageJSON object and determine if it is an ESM or CJS
+// package, with some other convenience logic thrown in for good measure.
+//
+// Ref:
+// https://github.com/npmgraph/npmgraph/pull/326#issuecomment-2972640439
+function detectPackageType(pkg: PackageJSON) {
+  // If the package name starts with @types/, then it's TS-types only
+  if (pkg.name.startsWith?.('@types/')) {
+    return { esm: false, cjs: false, types: true };
+  }
+
+  const pkgType: PackageModuleType = {
+    // If pkg#type is not set to 'module', assume it's CJS
+    cjs: pkg.type !== 'module',
+
+    // If pkg#type is set to 'module', then it's ESM
+    esm: pkg.type === 'module',
+
+    // If pkg#types is set, then TS types are available
+    types: pkg.types !== undefined,
+  };
+
+  // Inspect package#main
+  if (pkg.main?.endsWith?.('.mjs')) pkgType.esm = true;
+  if (pkg.main?.endsWith?.('.cjs')) pkgType.cjs = true;
+
+  // Inspect package#exports (recursively)
+  return _detectExports(pkg.exports, pkgType);
+}
+
+// Loosely inspect package.json#exports for module type (recursive)
+function _detectExports(exports: unknown, pkgType: PackageModuleType) {
+  if (Array.isArray(exports)) exports.some(v => _detectExports(v, pkgType));
+
+  if (exports && typeof exports === 'object') {
+    for (const [k, v] of Object.entries(exports)) {
+      if (k === 'import') {
+        pkgType.esm = true;
+      } else if (k === 'require') {
+        pkgType.cjs = true;
+      } else if (k === 'default') {
+        pkgType.esm = true;
+        pkgType.cjs = true;
+      }
```

Comment:
this doesn't sound right, if a package were to have only the default export it would only have one of them, but it would show up as dual

Reply:
Fixed.  I read through the `node-modules.dev` and `npm-esm-vs-cjs` code a bit more closely and have tried to follow their lead on this. (I'm not directly pulling in the code from those projects, however, as it doesn't quite do what I'm after.)

Reply:
FWIW: https://npmgraph-git-dualmodules-broofas-projects.vercel.app/?q=webdriver#color=moduleType&zoom=w seems to have a pretty good selection of module types... if you wanted to click through some of the `package.json` files for some of the modules to verify the detection logic works correctly.

(And, yes, I know there should be unit tests for this, but that's time and energy I don't have at the moment. :-) )